src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs

```diff
@@ -1448,14 +1448,18 @@ private bool WriteTouchedFiles(DiagnosticBag diagnostics, TouchedFileLogger? tou
 
         protected virtual ImmutableArray<AdditionalTextFile> ResolveAdditionalFilesFromArguments(List<DiagnosticInfo> diagnostics, CommonMessageProvider messageProvider, TouchedFileLogger? touchedFilesLogger)
         {
-            var builder = ImmutableArray.CreateBuilder<AdditionalTextFile>();
+            var builder = ArrayBuilder<AdditionalTextFile>.GetInstance();
+            var filePaths = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
```

Comment:
Assuming this is the fix for the correctness issue you were debugging. Can you elaborate a bit on this? Does this mean we were seeing the same additional file multiple times?

Reply:
@jaredpar  Yeah, specifically the `PublicAPI.[Un]Shipped.txt` files. They are bought in via our build and also added via the analyzer package. We were failing in our assertion that we succeed when adding to the set here https://github.com/dotnet/roslyn/blob/33bddf36090b2966f758f27d33b04814d8b429ac/src/Compilers/Core/Portable/SourceGeneration/Nodes/InputNode.cs#L44-L45 because we're trying to add the same item twice. Previously we didn't catch it because we were doing reference equality rather than content based. 

This is obviously a correctness issue: if we don't dedupe then a generator could end up generating the same tree twice for a given additional file.

Reply:
Where in the command line do we filter out the duplicate files?

Reply:
I was about to say you're commenting on it, then I looked at it and also wondered how it was working. Turns out I had a test bug too :( 

Now fixed. We expand the paths using the same logic added in https://github.com/dotnet/roslyn/pull/56588/ and then compare if they are the same.